_posts/2020-09-01-java-giants.md

```diff
@@ -0,0 +1,147 @@
+print (dataShared).  //never gonna happen 
+```
+2.  This rewriting is done by the JIT as it transforms bytecode to assembler at run time, if it so chooses to do this optimization.   If sharedDone is false at optimization time, we loop infinitely.  If it is true at optimization time, you would probably throw an exception repeatedly while trying to use uninitialized objects (because the Producer actually has not completed).
+3.  When stepping through the debugger, you are stepping over code in “bytecode land” only, not the rewritten assembler.   So, this infinite loop is impossible to reproduce in the debugger, and suddenly you are questioning your career choice.
+
+Terrifying?   Yes, yes I believe so.
+
+## Chris Newland
+Our mystery is solved, but the greater lesson here is that the JIT will rewrite your code under high loads for performance, and a "loose understanding" of java coding and memory-model rules can lead to unintended consequences.   A great (surprisingly short) article to see JIT rewriting in detail is Shipilev's first JVM Anatomy Park article [Lock Coursening and Loops](https://shipilev.net/jvm/anatomy-quarks/1-lock-coarsening-for-loops).   But, we can do even better.  
+
+A fourth giant related to this discussion is **Chris Newland** who created the visualization tool [JITWatch](https://github.com/AdoptOpenJDK/jitwatch) for you to see your code, the compiled bytecode, and the JIT assembler all in one glorious screen (and more!).   With this tool you could verify the infinite loop in the assembly.   Enjoy my favorite presentation of his on his tool called [Understanding HotSpot JVM Performance with JITWatch](http://vimeo.com/181925278).
+
+To really show how different assembly can be from the original, and as an excuse to play with JMH and JITWatch, suppose we have a simpler JMH test program with the following:
+```java
+    @Param("525600")
+    long answer;
+
+    static long howDoYouMeasure_MeasureAYear(
+            int daylights, int sunsets,
+            int midnights, long cupsOfCoffee,
+            int inches, int miles, long laughter, short strife){
+
+        int minutes = 525_600;  //FYI 0x80520 in hex
+        long result = daylights+sunsets+midnights+cupsOfCoffee;
+        result += inches + miles + +laughter + strife;
+        result -=result + minutes;
+        return result;
+    }
+
+    @Benchmark
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
+    public boolean get() {
+        long result = howDoYouMeasure_MeasureAYear(daylights, sunsets, midnights, cupsOfCoffee,
+                inches, miles, laughter, strife);
+        return result == answer;
+    }
+```
+This method always returns 525_600 regardless of input and convoluted logic.   The first time that the JIT compiles this to assembler, it will transform the method "as-is".   In JITWatch, we see this for the get() method: 
+![get_c1](/assets/images/2020-09-14-get-c1.png)
+Figure 2:  get() with bytecode and assembly.
```

Comment:
That one is hard to read, at least on my screen, on mobile might be also hard, can you somehow "magnify" this?